core/network-libp2p/src/custom_proto/handler.rs

```diff
@@ -347,14 +357,40 @@ where TMessage: CustomMessage, TSubstream: AsyncRead + AsyncWrite {
 				PerProtocolState::Poisoned
 			}
 
-			st @ PerProtocolState::Init(_) => {
+			PerProtocolState::Init { substreams, mut init_deadline } => {
+				match init_deadline.poll() {
+					Ok(Async::Ready(())) =>
+						error!(target: "sub-libp2p", "Handler initialization process is too long"),
+					Ok(Async::NotReady) => {}
+					Err(_) => error!(target: "sub-libp2p", "Tokio timer has errored")
+				}
+
 				return_value = None;
-				st
+				PerProtocolState::Init { substreams, init_deadline }
 			}
 
-			st @ PerProtocolState::Opening { .. } => {
-				return_value = None;
-				st
+			PerProtocolState::Opening { mut deadline } => {
+				match deadline.poll() {
+					Ok(Async::Ready(())) => {
+						deadline.reset(Instant::now() + Duration::from_secs(60));
+						let event = CustomProtosHandlerOut::ProtocolError {
+							protocol_id: self.protocol.id(),
+							is_severe: false,
+							error: "Timeout when opening protocol".to_string().into(),
+						};
+						return_value = Some(ProtocolsHandlerEvent::Custom(event));
+						PerProtocolState::Opening { deadline }
+					},
+					Ok(Async::NotReady) => {
+						return_value = None;
+						PerProtocolState::Opening { deadline }
+					},
+					Err(_) => {
+						error!(target: "sub-libp2p", "Tokio timer has errored");
+						return_value = None;
```

Comment:
Why isn't that the default `return_value`?

Reply:
So that the compiler detects if we forget to set the value of `return_value`.

Reply:
Good point^^